### PR TITLE
Add type hints and return types to methods in utils package

### PR DESCRIPTION
**Pull request by adrianfusco**
- Add type hints for utils subpackage
- Remove type hints in docstring

Note: It seems that we aren't able to specify the return type for a self class if Python version <= 3.11 (from typing import Self - https://peps.python.org/pep-0673/). I've seen the way to specify the same class is in a string way. If the self class is e.g. Production then the type should be 'Production'.

### CODE REVIEW

File: cibyl/utils/net.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class DownloadError(Exception):
     """
 
 
-def download_file(url, dest):
+def download_file(url: str, dest: str):
```
Comment (jsanemet) [inline cibyl/utils/net.py:29]: Returns '-> None'

File: cibyl/utils/net.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ def download_file(url, dest):
                 file.write(chunk)
 
 
-def download_into_memory(url, session=None):
+def download_into_memory(url: str, session: requests.Session = None) -> str:
```
Comment (jsanemet) [inline cibyl/utils/net.py:72]: Mark 'session' with 'Optional'.

File: cibyl/utils/dicts.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 LOG = logging.getLogger(__name__)
 
 
-def subset(dictionary, keys):
+def subset(dictionary, keys: list) -> dict:
```
Comment (cescgina) [inline cibyl/utils/dicts.py:23]: dictionary is missing the annotation, a general `dict` one should be fine for this function, I believe

File: cibyl/utils/sorting.py [hunk 4]
```diff
@@ -86,8 +81,8 @@ class BubbleSortAlgorithm(SortingAlgorithm):
     """
 
     @overrides
-    def sort(self, iterable):
-        def compare(lft, rght):
+    def sort(self, iterable: Iterable) -> list:
+        def compare(lft: Any, rght: Any):
```
Comment (jsanemet) [inline cibyl/utils/sorting.py:85]: We should know when to be less demanding. Type hinting an inner function does not have much value, even less if the types are undefined. Nobody is going to be able to see it from the outside.

**Review by jsanemet — CHANGES_REQUESTED** [review]
I still have a hard time seeing the value of redoing something that was already done. More so when that redo could add mistakes.

On any case, I have added some comments on a few places I think need to be revisited.

File: cibyl/utils/files.py [hunk 5]
```diff
@@ -25,49 +26,44 @@ class FileSearch:
...
         return self
 
-    def with_excluded(self, excluded):
+    def with_excluded(self, excluded: list) -> 'FileSearch':
         """Limits the search to files that are not in the excluded list. If this
         is called more than once, then the filters are joined together
         following an 'OR' approach.
 
         :param excluded: The file names to filter by.
         :type excluded: list
```
Comment (cescgina) [inline cibyl/utils/files.py:65]: This should be removed also

File: cibyl/utils/files.py [hunk 6]
```diff
@@ -77,10 +73,9 @@ def get(self):
         found by the search.
 
         :return: Paths to the files.
-        :rtype: list[str]
```
Comment (cescgina) [inline cibyl/utils/files.py:80]: This type hint is removed but not added to the function

File: cibyl/utils/filtering.py [hunk 7]
```diff
@@ -39,61 +39,52 @@
...
 
     :param model: model information obtained from jenkins
-    :type model: str
     :param user_input: input argument specified by the user
-    :type model_urls: :class:`.Argument`
     :param field_to_check: Job field to perform the check
-    :type field_to_check: str
     :param default_user_value: Default value to use if the user input contains
     no value
     :type default_user_value: list
```
Comment (cescgina) [inline cibyl/utils/filtering.py:86]: Forgot to remove this line

File: cibyl/utils/filtering.py [hunk 8]
```diff
@@ -39,61 +39,52 @@
...
     :returns: Whether the model satisfies user input
-    :rtype: bool
     """
     return model[field_to_check] in user_input.value
 
 
 def satisfy_case_insensitive_match(model: Dict[str, str], user_input: Argument,
                                    field_to_check: str,
-                                   default_user_value: List[str] = None):
+                                   default_user_value: List[str] = None
```
Comment (cescgina) [inline cibyl/utils/filtering.py:74]: This should be wrapped with Optional

File: cibyl/utils/reflection.py [hunk 9]
```diff
@@ -38,19 +38,18 @@ def load_module(path):
     return module
 
 
-def get_classes_in(__module, predicate=None, return_name=False):
+def get_classes_in(
+        __module: ModuleType,
+        predicate: Callable = None,
```
Comment (cescgina) [inline cibyl/utils/reflection.py:43]: Type hint should be Optional[Callable]

File: cibyl/utils/strings.py [hunk 10]
```diff
@@ -57,18 +55,17 @@ def level(self):
             """
             return self._level
 
-        def append(self, text):
+        def append(self, text: str):
```
Comment (cescgina) [inline cibyl/utils/strings.py:58]: Missing return value of None

File: cibyl/utils/dicts.py (hunk 3 above)
Comment (adrianfusco) [reply cibyl/utils/dicts.py]: Changed. Thanks!

File: cibyl/utils/files.py [hunk 11]
```diff
@@ -25,49 +26,44 @@ class FileSearch:
...
         :param directory: Path to directory to look for files in.
-        :type directory: str
         """
         self._directory = directory
         self._recursive = False
         self._extensions = []
         self._excluded = []
 
-    def with_recursion(self):
+    def with_recursion(self) -> 'FileSearch':
```
[the remark this answers is not in the document]
Comment (adrianfusco) [reply cibyl/utils/files.py:39]: As we discussed in chat it seems that there's not another way to do it. I added a note in the commit message. Not available for self classes if we're using Python < 3.11

File: cibyl/utils/files.py [hunk 12]
```diff
@@ -123,27 +116,25 @@ def _copy_for(self, directory):
...
 
-def get_first_available_file(filenames, file_check=is_file_available):
+def get_first_available_file(filenames: list,
+                             file_check: bool = is_file_available
+                             ) -> Union[bytes, str, None]:
     """Searches for the first file out of the provided paths that exists
     on the host's drive.
 
     :param filenames: A list of paths pointing to different files.
-    :type filenames: :class:`typing.Iterable[bytes | str | PathLike]`
```
[the remark this answers is not in the document]
Comment (adrianfusco) [reply cibyl/utils/files.py:142]: Changed. Thanks!

File: cibyl/utils/filtering.py [hunk 13]
```diff
@@ -39,61 +39,52 @@
...
     """Check whether model should be included according to the user input. The
     model should be added if the information provided field_to_check
     (the model name or url for example) is present in the user_input values.
 
     :param model: model information obtained from jenkins
-    :type model: str
     :param user_input: input argument specified by the user
-    :type model_urls: :class:`.Argument`
     :param field_to_check: Job field to perform the check
     :param field_to_check: str
```
[the remark this answers is not in the document]
Comment (adrianfusco) [reply cibyl/utils/filtering.py]: Changed. Thanks!

File: cibyl/utils/logger.py [hunk 14]
```diff
@@ -59,15 +56,13 @@ def configure_file_logging(log_file, level):
...
 
 
-def configure_logging(log_mode, log_file, level=logging.INFO):
+def configure_logging(log_mode: str, log_file: str, level: int = logging.INFO):
     """Configure logging format and level.
 
     :param log_mode: Where to send the logs, file, terminal or both
-    :type log_mode: str
     :param log_file: Path to log file
     :param log_file: str
```
[the remark this answers is not in the document]
Comment (adrianfusco) [reply cibyl/utils/logger.py:67]: Changed. Thanks!

File: cibyl/utils/net.py (hunk 2 above)
Comment (adrianfusco) [reply cibyl/utils/net.py]: Changed. Thanks!

File: cibyl/utils/files.py (hunk 5 above)
Comment (adrianfusco) [reply cibyl/utils/files.py:65]: Changed. Thanks!

File: cibyl/utils/filtering.py (hunk 7 above)
Comment (adrianfusco) [reply cibyl/utils/filtering.py]: Changed. Thanks!

File: cibyl/utils/filtering.py (hunk 8 above)
Comment (adrianfusco) [reply cibyl/utils/filtering.py]: Changed. Thanks!

File: cibyl/utils/strings.py (hunk 10 above)
Comment (adrianfusco) [reply cibyl/utils/strings.py]: Changed. Thanks!

File: cibyl/utils/filtering.py [hunk 15]
```diff
@@ -151,29 +136,25 @@ def filter_topology(model: Dict[str, str], operator: str, value: str,
...
-    :rtype: bool
     """
     try:
         return bool(re.search(pattern, string))
     except sre_constants.error:
         return False  # Do not crash against invalid patterns
 
 
-def apply_filters(iterable, *filters):
+def apply_filters(iterable: Iterable, *filters: list) -> list:
```
[the remark this answers is not in the document]
Comment (adrianfusco) [reply cibyl/utils/filtering.py]: Changed. Thanks!

File: cibyl/utils/dicts.py [hunk 16]
```diff
@@ -47,7 +44,7 @@ def subset(dictionary, keys):
     return result
 
 
-def nsubset(dictionary, keys):
+def nsubset(dictionary, keys: list) -> dict:
```
Comment (cescgina) [inline cibyl/utils/dicts.py:47]: This function is still missing the type hint for dictionary, and has it annotated in the docstring

File: cibyl/utils/files.py (hunk 5 above)
Comment (cescgina) [reply cibyl/utils/files.py:65]: I'm still seeing the line, maybe you forgot to push some commit?

File: cibyl/utils/filtering.py [hunk 17]
```diff
@@ -127,21 +112,18 @@ def satisfy_range_match(model: Dict[str, str], user_input: Argument,
...
     """Check whether model should be included according to the user input. The
     model should be added if the its topology is consistent with the components
     requested by the user (number of controller or compute nodes).
 
     :param model: model information obtained from jenkins
-    :type model: str
     :param operator: operator to filter the topology with
-    :type operator: str
     :param value: Value to use in the comparison
     :param value: str
```
Comment (cescgina) [inline cibyl/utils/filtering.py:123]: I just realised this lines should be removed, but were missing the type keyword before.

File: cibyl/utils/dicts.py (hunk 16 above)
Comment (cescgina) [reply cibyl/utils/dicts.py:47]: Sure, I agree that there is no point in specifying further than dict, but in this function it's still in the docstring and not yet in the function signature for `dictionary` argument. It's nitpicking, but just for consistency

File: cibyl/utils/filtering.py (hunk 17 above)
Comment (adrianfusco) [reply cibyl/utils/filtering.py:123]: Changed. Thanks!

**Comment by jsanemet** [comment]
With @adrianfusco on PTO, we will need to decide who concludes work on this PR.

**Comment by cescgina** [comment]
I can do it, are there any pending changes, @jsanemet ?

**Comment by jsanemet** [comment]
There are a couple of conversations still open. If you consider everything done, we can close this one down.

**Comment by cescgina** [comment]
Ok, let me do a couple quick chages for that and we can merge it.

**Comment by cescgina** [comment]
Done @jsanemet, take a look and if it ok by you, we can merge it

**Review by jsanemet — APPROVED** [review]
Looks good.

**Pull request merged**